core/base/compactTriangulation/CompactTriangulation.h

```diff
@@ -0,0 +1,4214 @@
+/// \sa ttk::CompactTriangulationPreconditioning
+
+#pragma once
+
+// base code includes
+#include <AbstractTriangulation.h>
+#include <CellArray.h>
+#include <FlatJaggedArray.h>
+#include <algorithm>
+#include <boost/functional/hash.hpp>
+#include <boost/unordered_map.hpp>
+#include <boost/unordered_set.hpp>
+#include <list>
+
+#define EDGE_ID 1
+#define TRIANGLE_ID 2
+
+using namespace std;
+namespace ttk {
+
+  class ImplicitCluster {
+  private:
+    /* components */
+    SimplexId nid;
+    vector<array<SimplexId, 2>> *internalEdgeList_;
+    vector<array<SimplexId, 3>> *internalTriangleList_;
+    boost::unordered_map<array<SimplexId, 2>, SimplexId> *internalEdgeMap_;
+    boost::unordered_map<array<SimplexId, 2>, SimplexId> *externalEdgeMap_;
+    boost::unordered_map<array<SimplexId, 3>, SimplexId> *internalTriangleMap_;
+    boost::unordered_map<array<SimplexId, 3>, SimplexId> *externalTriangleMap_;
+    /* boundary cells */
+    vector<bool> *boundaryVertices_;
+    vector<bool> *boundaryEdges_;
+    vector<bool> *boundaryTriangles_;
+    /* vertex relationships */
+    FlatJaggedArray *vertexEdges_;
+    FlatJaggedArray *vertexLinks_;
+    FlatJaggedArray *vertexNeighbors_;
+    FlatJaggedArray *vertexStars_;
+    FlatJaggedArray *vertexTriangles_;
+    /* edge relationships */
+    // edgeVertex relation can be extracted from internal edge list
+    FlatJaggedArray *edgeLinks_;
+    FlatJaggedArray *edgeStars_;
+    FlatJaggedArray *edgeTriangles_;
+    /* triangle relationships */
+    // triangleVertex relation can be extracted from internal triangle list
+    vector<array<SimplexId, 3>> *triangleEdges_;
+    FlatJaggedArray *triangleLinks_;
+    FlatJaggedArray *triangleStars_;
+    /* cell relationships */
+    vector<array<SimplexId, 6>> *tetraEdges_;
+    FlatJaggedArray *cellNeighbors_;
+    vector<array<SimplexId, 4>> *tetraTriangles_;
+
+  public:
+    ImplicitCluster(SimplexId id) {
+      /* components */
+      nid = id;
+      internalEdgeList_ = nullptr;
+      internalTriangleList_ = nullptr;
+      internalEdgeMap_ = nullptr;
+      externalEdgeMap_ = nullptr;
+      internalTriangleMap_ = nullptr;
+      externalTriangleMap_ = nullptr;
+      /* boundary cells */
+      boundaryEdges_ = nullptr;
+      boundaryVertices_ = nullptr;
+      boundaryTriangles_ = nullptr;
+      /* vertex relationships */
+      vertexEdges_ = nullptr;
+      vertexLinks_ = nullptr;
+      vertexNeighbors_ = nullptr;
+      vertexStars_ = nullptr;
+      vertexTriangles_ = nullptr;
+      /* edge relationships */
+      edgeLinks_ = nullptr;
+      edgeStars_ = nullptr;
+      edgeTriangles_ = nullptr;
+      /* triangle relationships */
+      triangleLinks_ = nullptr;
+      triangleEdges_ = nullptr;
+      triangleStars_ = nullptr;
+      /* cell relationships */
+      tetraEdges_ = nullptr;
+      cellNeighbors_ = nullptr;
+      tetraTriangles_ = nullptr;
+    }
+    ~ImplicitCluster() {
+      delete internalEdgeList_;
```

Comment:
we strongly discourage the explicit use of `new`/`delete` (which are now usually regarded as non-recommendable programming practices), unless there's a very good reason.
in this example, it seems that `internalEdgeList_` could simply be declared as `vector<array<SimplexId, 2>>` instead of `vector<array<SimplexId, 2>> *` (pointer).
is there any motivation for using pointers here?

Reply:
Previously I thought it might be helpful to use dynamic allocation for relation arrays (like removing the unused ones to release the allocated memory), but the implementation does not work like that. I will update the code accordingly.

Reply:
solved